### PR TITLE
[#15735][feat] KV-cache transfer/staging between two TRT-LLM instances

### PR DESCRIPTION
**Pull request by nafis271**
## Description

Implements #15735.

Builds on the disagg transceiver to add a general, **cross-instance** KV-transfer mechanism. A
`generation_only` request with a new **`kv_transfer_only`** flag pulls the prompt's KV from a
**source instance** over the transceiver, commits it to the local reuse cache, and **finishes
immediately, producing no output tokens**. The two instances just need the cache transceiver enabled — they
are not limited to a dedicated context/generation pair, and need not be co-located (UCX does
cross-node over IB too).

### New API

**`POST /v1/kv_transfer`** — accepts the request, starts the transfer in the background, and returns
`202` immediately (does not wait for the transfer to finish).
- Request body:
  - `source` (required) — base URL of the instance to pull the prompt's KV from.
  - `messages` **or** `prompt` (one required) — the prompt, in chat or completions form.
  - `model` (optional) — defaults to the server's served model.
- Response: `202` with `{status, request_hash}`, where `status` is one of `accepted`, `deduped`
  (an identical prompt was seen within the dedup TTL), or `skipped_busy` (the in-flight transfer cap
  was reached). Errors: `400` (missing `source` or prompt), `403` (no allowlist configured, or
  `source` not on it).
- Guards: de-duplicates identical prompts seen within a TTL; caps concurrent transfers and sheds the
  excess as `skipped_busy`; and **requires** a `source` allowlist — the endpoint fails closed (403)
  until `KV_TRANSFER_SOURCE_ALLOW` is set, to avoid an SSRF primitive. Tunables:
  `KV_TRANSFER_MAX_INFLIGHT`, `KV_TRANSFER_DEDUP_TTL`, `KV_TRANSFER_SOURCE_ALLOW`.

**`GET /v1/kv_transfer/stats`** — observability counters: `fired`, `transferred`, `failed`,
`deduped`, `skipped_busy`, plus live `inflight`, `seen` (dedup-table size), and `max_inflight`.

### How it works

It composes the existing disaggregated-serving interface rather than adding new transport. The
endpoint runs the two-phase disagg flow internally: the `context_only` leg makes `source` prefill the
prompt and return its disagg ticket (the opaque transceiver handle for that prompt's KV), and
replaying that ticket back to this instance as a `generation_only` request (with `kv_transfer_only`)
makes the engine pull the KV over the transceiver and commit it for reuse. The internal
`generation_only` leg is issued over loopback HTTP so it goes through the normal serving path
(validation, `disaggregated_params` handling, executor submission).

### Changes (all additive; regular disagg/serving paths unchanged)

- **`disaggregated_params.py` / `serve/openai_protocol.py`** — add the `kv_transfer_only` field and
  pass it through the request converter.
- **`_torch/pyexecutor/py_executor.py`**
  - `_prepare_disagg_gen_transmission_complete`: when `kv_transfer_only`, commit the KV and
    `finish_by(LENGTH, 0)` after the normal sampler setup (skipping that setup corrupts the batched
    sampler under concurrency), then drop the request from the forward batch to skip a wasted
    decode — but only when other work remains; a lone transfer keeps its (output-discarded) forward,
    which is what finalizes the reuse commit (and avoids an empty batch).
  - `_handle_first_token_response`: skip already-finished requests. A `kv_transfer_only` req is
    finished at transmission-complete, so without this it would emit a first-token response *and* the
    terminal response — two `is_final` results for one `client_id`. The skip mirrors the existing
    `is_finished` guard in `_emit_first_token_responses`, and is a no-op for stock requests (which
    are never finished at first-token).
- **`serve/openai_server.py`** — the `POST /v1/kv_transfer` and `GET /v1/kv_transfer/stats` handlers.
  The endpoint fails closed without an allowlist (SSRF guard), and `_do_kv_transfer` checks both POST
  responses before counting a transfer, clears the dedup entry on failure (so transient errors can
  retry), and narrows its `except` to HTTP/JSON/timeout.

`kv_transfer_only` requests compose with the existing idle-only transfer-progress blocking, so they
pull and commit without stalling live traffic.

## Test Coverage

Adds **`tests/unittest/disaggregated/test_kv_transfer_endpoint.py`** — unit tests for the endpoint's
request handling (bad JSON / missing fields → 400, allowlist + fail-closed → 403, accept + counters,
dedup within TTL, in-flight cap → `skipped_busy`, and the `/stats` counters), with the background
transfer stubbed so no engine/GPU is required (10 tests).

The engine-level transfer + reuse path was validated manually on the rc19 release image with
gpt-oss-120b across two instances (one source, one destination): reuse correctness (the transferred KV
is reused on a subsequent request with correct output), zero cross-contamination under sustained mixed
load (concurrent unrelated requests + mid-stream cancellations), and no fatal engine errors. Happy to
add an automated 2-GPU disagg integration test following the `tests/unittest/disaggregated` patterns
if maintainers prefer.

(Note vs rc19: rc19 still has an always-on transfer-progress wait, so there `kv_transfer_only` also
needs excluding from that wait; `main` blocks only when idle, so this PR does not touch the blocking
path.)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work)).
- If PR introduces API changes, an appropriate PR label is added — this PR is **api-compatible** (the flag and endpoints are additive and off by default).
- Any new dependencies have been scanned for license and vulnerabilities — no new dependencies.
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes — n/a.
- Documentation updated as needed.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

@coderabbitai summary

### CODE REVIEW

File: tensorrt_llm/serve/openai_server.py [hunk 1]
```diff
@@ -631,6 +648,124 @@ def _check_health(self) -> bool:
...
+                    dp, request_type="generation_only", kv_transfer_only=True)
+                async with session.post(f"http://127.0.0.1:{self.port}{path}",
+                                        json=gen_body) as r2:
+                    await r2.read()
+            self._kvt_stats["transferred"] += 1
+        except Exception as e:
+            self._kvt_stats["failed"] += 1
+            logger.warning(f"[kv-transfer] {h[:8]} failed: {repr(e)[:120]}")
+        finally:
+            self._kvt_inflight -= 1
```
[the remark this answers is not in the document]
Comment (nafis271) [reply tensorrt_llm/serve/openai_server.py:783]: Addressed in latest commit

File: tensorrt_llm/serve/openai_server.py [hunk 2]
```diff
@@ -202,6 +204,21 @@ def __init__(
...
+            "skipped_busy": 0
+        }
+        self._kvt_seen = {}
+        self._kvt_inflight = 0
+        self._kvt_max = int(os.environ.get("KV_TRANSFER_MAX_INFLIGHT", "4"))
+        self._kvt_dedup_ttl = float(
+            os.environ.get("KV_TRANSFER_DEDUP_TTL", "300"))
+        _allow = os.environ.get("KV_TRANSFER_SOURCE_ALLOW", "").strip()
+        self._kvt_allow = set(s.strip() for s in _allow.split(",")
+                              if s.strip()) if _allow else None
```
[the remark this answers is not in the document]
Comment (nafis271) [reply tensorrt_llm/serve/openai_server.py:221]: Addressed in latest commit

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -4728,6 +4728,12 @@ def _prepare_disagg_gen_transmission_complete(self, scheduled_batch):
 
                 self._maybe_prepend_logprobs_and_logits(req, beam_width)
 
+                # kv_transfer_only: transfer done + first token adopted; finish now
+                # (no generation forward pass), KV is committed for reuse.
+                if (req.py_disaggregated_params is not None and getattr(
+                        req.py_disaggregated_params, "kv_transfer_only", None)):
+                    req.finish_by(FinishReason.LENGTH, 0)
```
[the remark this answers is not in the document]
Comment (nafis271) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Addressed in latest commit; it skips the forward when there is other work in the batch, and keeps the request when it is the only thing scheduled. I verified that dropping a lone transfer's forward loses its KV-reuse commit (a later query stops hitting it) and also leaves an empty batch that trips the overlap loop, so keeping it there (the decode is free while idle) avoids both.

**Comment by Tabrizian** [comment]
@nafis271 We are working on this proposal in this Github PR: https://github.com/NVIDIA/TensorRT-LLM/pull/13055 Is it possible to take a look at it and let us know if you have any feedback? Thanks.
